manim/utils/units.py

```diff
@@ -0,0 +1,86 @@
+import copy
+import logging
+from functools import wraps
+
+import numpy as np
+
+from ..config import config
+
+
+def _size_from_dimension(dim=0, pixels=False):
+    if dim == 0:
+        return config["frame_width"] if not pixels else config["pixel_width"]
+    elif dim == 1:
+        return config["frame_height"] if not pixels else config["pixel_height"]
+    else:
+        # TODO: determine size in z-direction
+        logging.getLogger("manim").warning(
+            f"Z-direction is not supported for units - returning size of x-dimension."
+        )
+        return config["frame_width"] if not pixels else config["pixel_width"]
+
+
+def _is_dimensional(value):
+    return isinstance(value, np.ndarray) and value.shape[0] <= 3
+
+
+class Unit:
+    def __init__(self, value):
+        self._value = value
+        self._converter = lambda val, dim: val
+
+    @property
+    def value(self):
+        return copy.deepcopy(self._value)
```

Comment:
To avoid issues where `_value` was being changed (which caused issues when `convert()` is called multiple times), I decided to copy the value. Are the cleaner methods of doing so?

Reply:
Can you post here an example where this is happening?

Reply:
Take the following example (inside some animation):
```python
unit = PercentUnit(10 * UP)

self.play(obj1.shift, unit) # obj1 is shifted by 10%
self.play(obj2.shift, unit) # obj2 is shifted by the amount obj1 is shifted (say 0.8 munits), but interpretated as %, so 0.8% instead of 10%
```

This is caused by the fact that we assign `return_value = self.value`. The return value is then modified which modifies the original value as well. Using `copy.deepcopy` avoids this issue.

Reply:
OK I see thanks.

How about you set the value on initialization and then never change it again? So treat each Unit instance as a constant.

Reply:
Do you mean converting it upon creation?

In that case say we have `unit = PercentageUnit(10)`. This value might be used by methods that apply to different dimensions:
```python
self.play(obj.shift_up, unit)
self.play(obj.shift_left, unit)
```
Here the 10% applies to different values (and thus gets a different value upon conversion). Hence we can only determine what value it is converted to when the unit is used. We could implement something where, if the dimension is known beforehand, we immediately calculate it. However, I don't think this will contribute to clean and consistent code.

Reply:
I agree. But see my latest long comment below.